### PR TITLE
fix(platform): fix search component keyboard accessibility and screen reader issues

### PR DESCRIPTION
**Pull request by DeepakSap14**
#### Please provide a link to the associated issue.
Closes SAP/fundamental-ngx#5581
Closes SAP/fundamental-ngx#5276

#### Please provide a brief summary of this pull request.
This PR fixing following accessibility bugs
1. Announcement of search suggestion when any search input entered
2. closing suggestion on Esc button press. Selecting suggestion and closing suggestion on Tab button press.
3. announcement of unwanted character.
4. opening of menu on spacebar

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [NA] update `README.md`
- [NA] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by DeepakSap14** [comment]
suggestion is not announced when opened until any item has focus. working on that for fix.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (kavya-b) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:287]:
```suggestion
                break;
```

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (kavya-b) [inline libs/platform/src/lib/components/menu/menu-trigger.directive.ts:110]: is it not expected to also close the menu/select the item on spacebar? I see for `enter`, that on opening menu, if `enter` is pressed it selects the focused item; should this also happen for `spacebar` ?

Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu-trigger.directive.ts:109]: as per https://wiki.wdf.sap.corp/wiki/display/Acc/Menu, selection of menu item is only on Enter/Click not on spacebar.
but i see that Core and UI5 is also selecting menu-item on spacebar.
will check this with QA team.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: fixed

**Review by kavya-b — APPROVED** [review]
LGTM

File: libs/platform/src/lib/components/menu/menu-item.component.ts
Comment (Betrozov) [inline libs/platform/src/lib/components/menu/menu-item.component.ts:64]: @DeepakSap14 please use KeyUtil here

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Using Chrome screenreader, issue 1 from the linked issue - "On typing the letters, a list of search result is displayed but the information regarding the search result list is not read by
the screen reader." - this is still not happening, user must navigate to the list with arrow keys

**Comment by platon-rov** [comment]
Should be bug 

> The button read as ‘Submit’. Additional information is not conveyed.

Fixed in scope on current PR?

File: libs/platform/src/lib/components/menu/menu-item.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu-item.component.ts]: done

**Comment by DeepakSap14** [comment]
> Should be bug
> 
> > Submit button read just as ‘Submit’. Additional information is not conveyed.
> 
> Fixed in scope on current PR?

@platon-rov 
have included the fix in this PR.

**Comment by DeepakSap14** [comment]
> Using Chrome screenreader, issue 1 from the linked issue - "On typing the letters, a list of search result is displayed but the information regarding the search result list is not read by
> the screen reader." - this is still not happening, user must navigate to the list with arrow keys

@mikerodonnell89 
fixed "reading of search suggestion count".

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/menu/menu-trigger.directive.ts:108]: KeyUtil?

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/menu/menu-trigger.directive.ts:110]: KeyUtil?

File: libs/platform/src/lib/components/search-field/search-field.component.spec.ts
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/search-field/search-field.component.spec.ts:869]: Why is this commented out? If not used please remove it.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:221]: What if the user wants to customize this message?

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:481]: count++ maybe?

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (InnaAtanasova) [inline libs/platform/src/lib/components/search-field/search-field.component.scss:57]: Please reference the source of this block of code: https://webaim.org/techniques/css/invisiblecontent/

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (InnaAtanasova) [reply libs/platform/src/lib/components/search-field/search-field.component.scss:57]: I am pretty sure this will be needed in future again and is not related to any component specifically, so I suggest you talk to Manju and spend some extra time creating a separate component in css that is not exposed in the documentation. You then add this styling there and in ngx we can have a custom directive or component that will handle all the cases where we need an html element for a11y purposes.

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu-trigger.directive.ts]: using KeyUtil now.

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu-trigger.directive.ts]: using KeyUtil now.

File: libs/platform/src/lib/components/search-field/search-field.component.spec.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.spec.ts]: removed

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: added input property to get the message.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: done

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.scss:58]: raised https://github.com/SAP/fundamental-ngx/issues/5710 to track this.
will modify this, once directive is created.

reference provided.

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (platon-rov) [inline libs/platform/src/lib/components/menu/menu-trigger.directive.ts:108]: Somewhere it's named `event` and somewhere else `$event`, as far as I know, we're not using `$` in such cases. Should it be unified to the `event`?

File: libs/platform/src/lib/components/menu/menu.component.ts
Comment (platon-rov) [inline libs/platform/src/lib/components/menu/menu.component.ts:137]: https://github.com/SAP/fundamental-ngx/pull/5673/files#r655965061

File: libs/platform/src/lib/components/menu/menu-trigger.directive.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu-trigger.directive.ts]: made it uniform with other components. using `event`, removed `$event`.

File: libs/platform/src/lib/components/menu/menu.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/menu/menu.component.ts]: made it uniform with other components. using `event`, removed `$event`.

**Comment by DeepakSap14** [comment]
@mikerodonnell89 @InnaAtanasova 
your change request are answered. please re-review.

**Review by sKudum — APPROVED** [review]

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (InnaAtanasova) [reply libs/platform/src/lib/components/search-field/search-field.component.scss:59]: I am sorry @DeepakSap14 , this is no longer allowed in Fundamental-ngx. We can't approve anything that says "will be refactored once something else is done". Mark the PR as blocked and the review will continue once the issue 5710 is merged.

**Comment by DeepakSap14** [comment]
Closing this PR, until issue https://github.com/SAP/fundamental-ngx/issues/5710 is resolved and available as directive.

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.scss]: @InnaAtanasova using liveAnnouncer from cdk to read the search result. please review.
will raise a PR to remove `off-screen` component from fundamental-styles.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (platon-rov) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:229]: Should that field have any access modifier? And if it's not a part of public API it should starts with `_`

Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: now it is not getting used in template, made it private and included `_` in variable name.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (aberikashvili) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:298]: Isn't it better to move this kind of stuff to enum?

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (aberikashvili) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:494]: I'd recommend to place empty line before `return` for better readability

File: libs/platform/src/lib/components/search-field/search-field.component.html
Comment (aberikashvili) [inline libs/platform/src/lib/components/search-field/search-field.component.html:111]: Nice to see event specificity here

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (aberikashvili) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:311]: This could be simplified:

```suggestion
        setTimeout(() => this._currentSearchSuggestionAnnoucementMessage = '');
```

File: libs/platform/src/lib/components/search-field/search-field.module.ts
Comment (aberikashvili) [inline libs/platform/src/lib/components/search-field/search-field.module.ts:8]: Nice to see using sub packaging here

**Review by aberikashvili — COMMENTED** [review]
Please take a look at my comments

File: libs/platform/src/lib/components/search-field/search-field.component.scss
Comment (InnaAtanasova) [reply libs/platform/src/lib/components/search-field/search-field.component.scss]: You are missing approval from Platform team.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: fixed

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts:489]: fixed

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: using keyUtil from core and Keycodes from cdk now.

**Comment by DeepakSap14** [comment]
> Please take a look at my comments

@aberikashvili, your comments are addressed. please re-review once.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (kavya-b) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:164]:
```suggestion
    @Input()
    ariaLabelledby: string;
```
and in lines 169 and 175 too.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (kavya-b) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:255]:
```suggestion
        private readonly _elementRef: ElementRef,
```
?

**Review by sKudum — APPROVED** [review]

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: fixed formatting for all.

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: fixed

**Review by mikerodonnell89 — APPROVED** [review]
Code looks good and this works well.  One funny thing with the chrome screenreader is it is breaking up the result words by the bold letter - for example typing 'a' and navigating to 'Grape' the screenreader will say "Gee Arr Ape".  This can be fixed by using a `<b>` element rather than `<strong>` for the emphasized character.  I'll open another ticket for that because it applies to all components with this functionality

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (Betrozov) [inline libs/platform/src/lib/components/search-field/search-field.component.ts:495]: should we have to unsubscribe for dropdownValues$?

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/search-field/search-field.component.ts
Comment (DeepakSap14) [reply libs/platform/src/lib/components/search-field/search-field.component.ts]: fixed

**Review by aberikashvili — APPROVED** [review]
LGTM

**Review by Betrozov — APPROVED** [review]

**Comment by puru-hk** [comment]
**Issues 5276: Fixed** 


**Issues: 5581: Reopen**

Issue 1: Fixed

Issue 2: Fixed

Issue 3: 
Same issue persists.
OB: On activating the 'Clear' button, additional information 'A' is read.
EB: ideally, such additional information should be removed.
image

issue: 4 Fixed

JAWS 2021
URL: https://deploy-preview-5673--fundamental-ngx.netlify.app/#/platform/search-field

**Comment by DeepakSap14** [comment]
@puru-hk
about Issue 3, that extra information is getting read.
usecase is : we are typing `app` and then clearing it. so 'a' which is first character of 'app', that is getting cleared at last, and screen reader reads it.
if you type some other word like `basketball`, then on clear 'b' will be read.
Do you think that is an issue? i think it is the correct behaviour for a screen reader.

**Comment by DeepakSap14** [comment]
@puru-hk
Issue 3: this is not an issue from library side. It is default behaviour of Jaws screen reader. I did try to over-ride it, but it can't be done. It is not fixable.
On clearing the input text like apple, last cleared character a is getting read.
please take decision, we have other issues fixed. You can raise a different issue for issue 3 , so other fixed issue can be merged.

**Comment by puru-hk** [comment]
For Issue3: Different issue had been raised #6107. Because all other issues are fixed. Hence approve this PR.

**Review by puru-hk — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Pull request merged**
